core/common/src/main/java/alluxio/metrics/MetricsSystem.java

```diff
@@ -39,225 +43,241 @@
-  private List<Source> mSources = new ArrayList<>();
-  private MetricRegistry mMetricRegistry = new MetricRegistry();
-  private MetricsConfig mMetricsConfig;
-  private boolean mRunning = false;
+  /**
+   * Start sinks specified in the configuration. This is an no-op if the sinks have already been
+   * started.
+   * Note: This has to be called after Alluxio configuration is initialized.
+   */
+  public static void startSinks() {
+    String metricsConfFile = Configuration.get(PropertyKey.METRICS_CONF_FILE);
+    if (metricsConfFile.isEmpty()) {
+      LOG.info("Metrics is not enabled.");
+      return;
+    }
+    MetricsConfig config = new MetricsConfig(metricsConfFile);
+    startSinksFromConfig(config);
+  }
 
   /**
-   * Gets the sinks.
+   * Start sinks from a given metrics configuration. This is made public for unit test.
    *
-   * @return a list of registered Sinks
+   * @param config the metrics config
    */
-  public List<Sink> getSinks() {
-    return mSinks;
+  public static synchronized void startSinksFromConfig(MetricsConfig config) {
+    if (sSinks != null) {
+      LOG.warn("Sinks have already been started.");
+      return;
+    }
+    sSinks = new ArrayList<>();
+    Map<String, Properties> sinkConfigs =
+        MetricsConfig.subProperties(config.getProperties(), SINK_REGEX);
+    for (Map.Entry<String, Properties> entry : sinkConfigs.entrySet()) {
+      String classPath = entry.getValue().getProperty("class");
+      if (classPath != null) {
+        try {
+          Sink sink =
+              (Sink) Class.forName(classPath).getConstructor(Properties.class, MetricRegistry.class)
+                  .newInstance(entry.getValue(), METRIC_REGISTRY);
+          sSinks.add(sink);
+        } catch (Exception e) {
+          LOG.error("Sink class {} cannot be instantiated", classPath, e);
+        }
+      }
+    }
   }
 
   /**
-   * Gets the sources. Used by unit tests only.
-   *
-   * @return a list of registered Sources
+   * Stop all the sinks.
    */
-  public List<Source> getSources() {
-    return mSources;
+  public static synchronized void stopSinks() {
+    if (sSinks != null) {
+      for (Sink sink : sSinks) {
+        sink.stop();
+      }
+    }
+    sSinks = null;
   }
 
   /**
-   * Checks if the poll period is smaller that the minimal poll period which is 1 second.
-   *
-   * @param pollUnit the polling unit
-   * @param pollPeriod the polling period
-   * @throws IllegalArgumentException if the polling period is invalid
+   * @return the number of sinks started
    */
-  public static void checkMinimalPollingPeriod(TimeUnit pollUnit, int pollPeriod)
-      throws IllegalArgumentException {
-    int period = (int) MINIMAL_POLL_UNIT.convert(pollPeriod, pollUnit);
-    if (period < MINIMAL_POLL_PERIOD) {
-      throw new IllegalArgumentException("Polling period " + pollPeriod + " " + pollUnit
-          + " is below than minimal polling period");
+  public static synchronized int getNumSinks() {
```

Comment:
I actually used an reentrant lock in an earlier version. I removed that because I don't think it helps readability. I have added a GuardedBy annotation on the mSinks field. I think that achieves what you want.